### ISSUE TITLE
`included` parameter potentially includes the object itself

### ISSUE THREAD

**Issue by puremath** [django-json-api/django-rest-framework-json-api#524]
Let's say we have a class `Room` and a class `Person` with a foreign key to `Room`. Now I want to create a new person in the room and reload all the people already in the room because adding the person might have changed them for some reason. Therefor I send a `POST` request containing the new person to 
```
persons/?include=room.persons`
```
This lead to the newly created person to be included twice, once in the `data` property of the reply and once in the `included` property. Even though this isn't explicitly forbidden by jsonapi.org, Ember doesn't like this. IMHO an object shouldn't be included twice in a reply - meaning that this should be fixed on the DJA side of things.

**Comment by puremath** [comment]
I didn't post this anywhere else. When I re-read the specification I thought it is pretty clear on this:

> A compound document MUST NOT include more than one resource object for each type and id pair.
> 
> 
> Note: In a single document, you can think of the type and id as a composite key that uniquely references resource objects in another part of the document.
>
> Note: This approach ensures that a single canonical resource object is returned with each response, even when the same resource is referenced multiple times.

**Comment by sliverc** [comment]
> In a compound document, all included resources MUST be represented as an array of resource objects in a top-level included member.

Here the specification states that compound document is the top-level included member. This doesn't include the default resource object, so the statement that `A compound document MUST NOT include more than one resource` is as I understand it only relevant for the included member.

If it is understood this way how DJA implements it now would be OK.

I think in terms of payload fixing this would actually make sense, I am just not so sure on the clarity of the specification on this point, so it would be good to clarify it.

**Comment by sliverc** [comment]
I have overread this part, but with this in mind I agree that a resource object should not be added twice including primary data.

I guess it never can hurt to double check though with JSON API. Marking this as bug and a Pull Request is very welcome.

**Comment by sliverc** [comment]
The resource itself should be part of the included_cache so it will be skipped in the renderer if it gets included again.

See https://github.com/django-json-api/django-rest-framework-json-api/blob/master/rest_framework_json_api/renderers.py#L383

**Comment by puremath** [comment]
I'm almost ready to submit the PR. Before I do a couple of questions:

* I wrote a integration test which tests this case but no unit test. Do we also need to unit-test this case?

* Adding the resource itself to the included_cache is IMHO not the best approach as we then have to removed it again before building the included attribute of the response. Therefor I decided to just check in the end whether it is in the cache and remove it in this case. Could this lead to any problems I didn't yet fully grasp?
